### DISCUSSION TITLE
Using VERCEL_URL shows deployment URL rather than assigned domain

### DISCUSSION THREAD

**Discussion by dextermb** [vercel/community#6429] [Help] [answered]
### Summary

Hey team

I'm working on a project which sends magic links to sign in. Due to this I need to read from the `VERCEL_URL` environment variable.

I've noticed that even though I have `*-preview.vercel.app` assigned to my `preview` branch that whenever I read from `VERCEL_URL` on that domain I get the deployment URL `*-<deploymentId>-<username>.vercel.app`.

I would expect `VERCEL_URL` to take the assigned domain over the deployment URL.

Any ideas how to solve this without manually overwriting the environment variable?

### Steps to Reproduce

1. Create a project
2. Connect to a GitHub repository
3. Assign domain to branch (such as `preview`)
4. Push changes to `preview`, and log `VERCEL_URL`

**Comment by dextermb** [comment]
This seems to happen for production deployments as well. Am I misunderstanding something here?

**Comment by amyegan** [comment]
Hey @dextermb. The `VERCEL_URL` value should be the the [generated deployment URL](https://vercel.com/docs/deployments/generated-urls), so this is expected behavior. 

If you are using predictable URLs for **production** and **preview**, then you may prefer to use `VERCEL_ENV` to choose the correct URL for the current environment. There is also a `VERCEL_BRANCH_URL` which displays the [generated Git branch URL](https://vercel.com/docs/deployments/generated-urls#url-with-git-branch). 

You can learn more about the available System Environment Variables at [vercel.com/docs/projects/environment-variables/system-environment-variables](https://vercel.com/docs/projects/environment-variables/system-environment-variables#)

**Reply by dextermb** [reply, marked as answer]
Hey Amy, thanks for responding.

It would be great to be able to determine a static URL from a Vercel provided ENV variable, for example a domain assigned to `main` or `preview`.

Ideally I wouldn't want to have two places to manage which domains are for what environments – Vercel dashboard and our own `.env` file.

For now however I'll use `VERCEL_ENV` to then determine in-code what domain to point to.

**Reply by depsimon** [reply]
Vercel is currently exposing 3 URLs on all environments:

- `VERCEL_URL`: the deployment URL
- `VERCEL_BRANCH_URL`: the branch preview URL
- `VERCEL_PROJECT_PRODUCTION_URL`: the shortest production URL

But those are identical for a deployment whether you consult the deployment URL or the branch URL.
While it makes sense, it'd be great if we could rely on a specific environment variable to get the visited URL. I need that in modules where I don't have access to the request object.

Anyway, my solution will be to always redirect to the `VERCEL_BRANCH_URL` in previews so that I'm sure it's consistent. It's too bad that we loose a Vercel feature for such a silly thing though.
